include/util/cell_storage.hpp

```diff
@@ -0,0 +1,295 @@
+
+        using RowIterator = WeightPtrT;
+        class ColumnIterator : public std::iterator<std::random_access_iterator_tag, EdgeWeight>
+        {
+          public:
+            explicit ColumnIterator(WeightPtrT begin, std::size_t row_length)
+                : current(begin), offset(row_length)
+            {
+            }
+
+            WeightRefT operator*() const { return *current; }
+
+            ColumnIterator &operator++()
+            {
+                current += offset;
+                return *this;
+            }
+
+            ColumnIterator &operator+=(int amount)
+            {
+                current += offset * amount;
+                return *this;
+            }
+
+            bool operator==(const ColumnIterator& other) const
+            {
+                return current == other.current;
+            }
+
+            bool operator!=(const ColumnIterator& other) const
+            {
+                return current != other.current;
+            }
+
+            int operator-(const ColumnIterator &other) const { return (current - other.current) / offset; }
+
+          private:
+            WeightPtrT current;
+            std::size_t offset;
+        };
+
+        std::size_t GetRow(NodeID node) const
+        {
+            return std::find(source_boundary, source_boundary + num_source_nodes, node) -
+                   source_boundary;
+        }
+        std::size_t GetColumn(NodeID node) const
+        {
+            return std::find(
+                       destination_boundary, destination_boundary + num_destination_nodes, node) -
+                   destination_boundary;
+        }
+
+      public:
+        std::pair<RowIterator, RowIterator> GetOutWeight(NodeID node) const
+        {
+            auto row = GetRow(node);
+            auto begin = weights + num_destination_nodes * row;
+            auto end = begin + num_destination_nodes;
+            return std::make_pair(begin, end);
+        }
+
+        std::pair<ColumnIterator, ColumnIterator> GetInWeight(NodeID node) const
+        {
+            auto column = GetColumn(node);
+            auto begin = ColumnIterator{weights + column, num_destination_nodes};
+            auto end =
+                ColumnIterator{weights + column + num_source_nodes * num_destination_nodes,
+                               num_destination_nodes};
+            return std::make_pair(begin, end);
+        }
+
+        auto GetSourceNodes() const
+        {
+            return std::make_pair(source_boundary, source_boundary + num_source_nodes);
+        }
+
+        auto GetDestinationNodes() const
+        {
+            return std::make_pair(destination_boundary,
+                                  destination_boundary + num_destination_nodes);
+        }
+
+        CellImpl(const CellData &data,
+                 WeightPtrT const all_weight,
+                 const NodeID *const all_sources,
+                 const NodeID *const all_destinations)
```

Comment:
Hm if you really want to user pointers at least assert them not being nullptrs. The interface is a bit ugly..